### ISSUE TITLE
Fixed Error

### ISSUE THREAD

**Issue by yxeGithub** [kridtapon/RSI-Optimization#1]
Fixed the error

`import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import yfinance as yf
from itertools import product
import seaborn as sns

def calculate_rsi(data, period=14):
    """
    Calculate the Relative Strength Index (RSI).
    """
    delta = data['Close'].diff(1)
    gain = delta.where(delta > 0, 0)
    loss = -delta.where(delta < 0, 0)

    avg_gain = gain.rolling(window=period).mean()
    avg_loss = loss.rolling(window=period).mean()

    rs = avg_gain / avg_loss
    rsi = 100 - (100 / (1 + rs))
    data['RSI'] = rsi
    return data

def calculate_drawdown(equity_curve):
    """
    Calculate drawdown and maximum drawdown from an equity curve.
    """
    rolling_max = equity_curve.cummax()
    drawdown = (equity_curve - rolling_max) / rolling_max
    max_drawdown = drawdown.min()
    return drawdown, max_drawdown

def backtest_rsi(data, rsi_period=14, overbought=70, oversold=30, stop_loss_pct=0.05, initial_balance=10000):
    """
    Backtest an RSI trading strategy.
    """
    data = calculate_rsi(data, period=rsi_period)
    balance = initial_balance
    position = 0
    entry_price = 0
    equity_curve = [initial_balance]  # Track balance over time

    for i in range(1, len(data)):
        row = data.iloc[i]
        prev_row = data.iloc[i - 1]

        # Buy signal: RSI crosses above the oversold level
        if position == 0 and prev_row['RSI'] < oversold and row['RSI'] >= oversold:
            position = balance / row['Close']
            entry_price = row['Close']
            balance = 0

        # Sell signal: RSI crosses below the overbought level or stop-loss triggered
        elif position > 0:
            stop_loss_price = entry_price * (1 - stop_loss_pct)
            if (row['RSI'] <= overbought and prev_row['RSI'] > overbought) or row['Close'] <= stop_loss_price:
                balance = position * row['Close']
                position = 0

        # Update equity curve
        equity_curve.append(balance if position == 0 else position * row['Close'])

    # Final balance if holding a position
    if position > 0:
        balance += position * data.iloc[-1]['Close']
        equity_curve.append(balance)

    # Calculate drawdown
    equity_curve = pd.Series(equity_curve)
    drawdown, max_drawdown = calculate_drawdown(equity_curve)

    return balance, max_drawdown, equity_curve

def optimize_rsi(data, rsi_periods, overbought_levels, oversold_levels, stop_loss_pcts):
    """
    Optimize RSI parameters.
    """
    results = []

    # Iterate over all parameter combinations
    for rsi_period, overbought, oversold, stop_loss_pct in product(rsi_periods, overbought_levels, oversold_levels, stop_loss_pcts):
        if overbought <= oversold:
            continue  # Skip invalid combinations

        final_balance, max_drawdown, _ = backtest_rsi(data, rsi_period, overbought, oversold, stop_loss_pct)
        results.append({
            'RSI Period': rsi_period,
            'Overbought Level': overbought,
            'Oversold Level': oversold,
            'Stop Loss %': stop_loss_pct,
            'Final Balance': final_balance,
            'Max Drawdown': max_drawdown
        })

    results_df = pd.DataFrame(results)
    return results_df

def plot_optimization_results(results_df):
    """
    Plot the optimization results.
    """
    pivot = results_df.pivot_table(
        index='RSI Period',
        columns='Overbought Level',
        values='Final Balance',
        aggfunc='max'
    )

    plt.figure(figsize=(10, 8))
    plt.title("Optimization Results: Final Balance")
    sns.heatmap(pivot, annot=True, fmt=".2f", cmap="viridis")
    plt.xlabel("Overbought Level")
    plt.ylabel("RSI Period")
    plt.show()

def plot_equity_curve(equity_curve):
    """
    Plot the equity curve.
    """
    plt.figure(figsize=(10, 6))
    plt.plot(equity_curve, label='Equity Curve', color='blue')
    plt.title("Equity Curve")
    plt.xlabel("Time")
    plt.ylabel("Equity Value")
    plt.legend()
    plt.grid()
    plt.show()

# Fetch stock data
ticker = "AAPL"  # Replace with your preferred stock ticker
data = df_N

# Optimize RSI parameters
rsi_periods = range(5, 21, 1)  # RSI periods from 5 to 20
overbought_levels = range(65, 86, 5)  # Overbought levels from 65 to 85
oversold_levels = range(15, 36, 5)  # Oversold levels from 15 to 35
stop_loss_pcts = [0.02, 0.05, 0.1]  # Stop-loss percentages

results_df = optimize_rsi(data, rsi_periods, overbought_levels, oversold_levels, stop_loss_pcts)

# Find the best parameters
best_params = results_df.loc[results_df['Final Balance'].idxmax()]
print("Best Parameters:")
print(best_params)

# Backtest with best parameters
_, _, best_equity_curve = backtest_rsi(
    data,
    rsi_period=int(best_params['RSI Period']),
    overbought=int(best_params['Overbought Level']),
    oversold=int(best_params['Oversold Level']),
    stop_loss_pct=float(best_params['Stop Loss %'])
)

# Plot equity curve
plot_equity_curve(best_equity_curve)

# Plot optimization results
plot_optimization_results(results_df)
`
